sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -180,6 +180,70 @@ static inline std::string GetHTTPMessagePreBody(Azure::Core::Http::Request const
 
   return httpRequest;
 }
+
+static void CleanupThread()
+{
+  using namespace Azure::Core::Http::_detail;
+  for (;;)
+  {
+    {
+      // Won't continue until the ConnectionPoolMutex is released from MoveConnectionBackToPool
+      std::unique_lock<std::mutex> lockForPoolCleaning(
+          CurlConnectionPool::g_curlConnectionPool.ConnectionPoolMutex);
+      // Wait the defined default time OR to the signal from the conditional variable.
+      // wait_until releases the mutex lock until it wakes up again or it's cancelled.
+      if (CurlConnectionPool::g_curlConnectionPool.ConditionalVariableForCleanThread.wait_for(
+              lockForPoolCleaning,
+              std::chrono::milliseconds(DefaultCleanerIntervalMilliseconds),
+              []() { return CurlConnectionPool::g_curlConnectionPool.ConnectionCounter == 0; }))
+      {
+        // Cancelled by another thead or no connections on wakeup
+        return;
+      }
+
+      // loop the connection pool index - Note: lock is re-taken for the mutex
+      for (auto index = CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.begin();
+           index != CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.end();
+           index++)
+      {
+        if (index->second.size() == 0)
+        {
+          // Move the next pool index
+          continue;
+        }
+
+        // Pool index with waiting connections. Loop the connection pool backwards until
+        // a connection that is not expired is found or until all connections are removed.
+        for (auto connection = index->second.end();;)
+        {
+          // loop starts at end(), go back to previous possition. We know the list is
+          // size() > 0 so we are safe to go end() - 1 and find the last element in the
+          // list
+          connection--;
+          if (connection->get()->IsExpired())
+          {
+            // remove connection from the pool and update the connection to the next one
+            // which is going to be list.end()
+            connection = index->second.erase(connection);
+            CurlConnectionPool::g_curlConnectionPool.ConnectionCounter -= 1;
```

Comment:
Do we really need this `ConnectionCounter`? What purpose does it serve other than requiring double book-keeping, one  for the map and one for the counter?

Reply:
The Pool is a dictionary of pools. Each connection-key (scheme, host, port, config etc) creates an index in the Pool where all the connections for that index will be keep. Because of that, counting the total connections in all indexes would require to traverse all indexes and sum in O(n),  That's why I introduced this extra state to get the total connections in O(1) just by paying a little extra sync between the variable-state and the pool updates.

Reply:
As far as I can tell, the only real place where you use `ConnectionCounter` is in the clean up thread method:
`[]() { return CurlConnectionPool::g_curlConnectionPool.ConnectionCounter == 0; }))`

Couldn't this be replaced with `g_curlConnectionPool.empty()` which is also O(1)?
https://www.cplusplus.com/reference/map/map/empty/

If `ConnectionCounter == 0` that means we either don't have any keys in the dictionary (either because we never had any or we removed them all), correct?